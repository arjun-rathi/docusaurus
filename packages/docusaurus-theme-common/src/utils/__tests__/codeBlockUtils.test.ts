/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  parseCodeBlockTitle,
  parseLanguage,
  parseLines,
} from '../codeBlockUtils';

describe('parseCodeBlockTitle', () => {
  it('parses double quote delimited title', () => {
    expect(parseCodeBlockTitle(`title="index.js"`)).toBe(`index.js`);
  });

  it('parses single quote delimited title', () => {
    expect(parseCodeBlockTitle(`title='index.js'`)).toBe(`index.js`);
  });

  it('does not parse mismatched quote delimiters', () => {
    expect(parseCodeBlockTitle(`title="index.js'`)).toBe(``);
  });

  it('parses undefined metastring', () => {
    expect(parseCodeBlockTitle(undefined)).toBe(``);
  });

  it('parses metastring with no title specified', () => {
    expect(parseCodeBlockTitle(`{1,2-3}`)).toBe(``);
  });

  it('parses with multiple metadata title first', () => {
    expect(parseCodeBlockTitle(`title="index.js" label="JavaScript"`)).toBe(
      `index.js`,
    );
  });

  it('parses with multiple metadata title last', () => {
    expect(parseCodeBlockTitle(`label="JavaScript" title="index.js"`)).toBe(
      `index.js`,
    );
  });

  it('parses double quotes when delimited by single quotes', () => {
    expect(parseCodeBlockTitle(`title='console.log("Hello, World!")'`)).toBe(
      `console.log("Hello, World!")`,
    );
  });

  it('parses single quotes when delimited by double quotes', () => {
    expect(parseCodeBlockTitle(`title="console.log('Hello, World!')"`)).toBe(
      `console.log('Hello, World!')`,
    );
  });
});

describe('parseLanguage', () => {
  it('works', () => {
    expect(parseLanguage('language-foo xxx yyy')).toBe('foo');
    expect(parseLanguage('xxxxx language-foo yyy')).toBe('foo');
    expect(parseLanguage('xx-language-foo yyyy')).toBeUndefined();
    expect(parseLanguage('xxx yyy zzz')).toBeUndefined();
  });
});

describe('parseLines', () => {
  it('does not parse content with metastring', () => {
    expect(parseLines('aaaaa\nnnnnn', '{1}', 'js')).toMatchSnapshot();
    expect(
      parseLines(
        `// highlight-next-line
aaaaa
bbbbb`,
        '{1}',
        'js',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `aaaaa
bbbbb`,
        '{1}',
      ),
    ).toMatchSnapshot();
  });
  it('does not parse content with no language', () => {
    expect(
      parseLines(
        `// highlight-next-line
aaaaa
bbbbb`,
        '',
        undefined,
      ),
    ).toMatchSnapshot();
  });
  it('removes lines correctly', () => {
    expect(
      parseLines(
        `// highlight-next-line
aaaaa
bbbbb`,
        '',
        'js',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `// highlight-start
aaaaa
// highlight-end
bbbbb`,
        '',
        'js',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `// highlight-start
// highlight-next-line
aaaaa
bbbbbbb
// highlight-next-line
// highlight-end
bbbbb`,
        '',
        'js',
      ),
    ).toMatchSnapshot();
  });
  it('respects language', () => {
    expect(
      parseLines(
        `# highlight-next-line
aaaaa
bbbbb`,
        '',
        'js',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `/* highlight-next-line */
aaaaa
bbbbb`,
        '',
        'py',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `// highlight-next-line
aaaa
/* highlight-next-line */
bbbbb
# highlight-next-line
ccccc
<!-- highlight-next-line -->
dddd`,
        '',
        'py',
      ),
    ).toMatchSnapshot();
    expect(
      parseLines(
        `// highlight-next-line
aaaa
/* highlight-next-line */
bbbbb
# highlight-next-line
ccccc
<!-- highlight-next-line -->
dddd`,
        '',
        '',
      ),
    ).toMatchSnapshot();
  });
});
