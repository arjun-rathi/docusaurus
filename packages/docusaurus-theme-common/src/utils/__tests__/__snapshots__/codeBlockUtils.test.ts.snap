// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLines does not parse content with metastring 1`] = `
{
  "code": "aaaaa
nnnnn",
  "highlightLines": [
    0,
  ],
}
`;

exports[`parseLines does not parse content with metastring 2`] = `
{
  "code": "// highlight-next-line
aaaaa
bbbbb",
  "highlightLines": [
    0,
  ],
}
`;

exports[`parseLines does not parse content with metastring 3`] = `
{
  "code": "aaaaa
bbbbb",
  "highlightLines": [
    0,
  ],
}
`;

exports[`parseLines does not parse content with no language 1`] = `
{
  "code": "// highlight-next-line
aaaaa
bbbbb",
  "highlightLines": [],
}
`;

exports[`parseLines removes lines correctly 1`] = `
{
  "code": "aaaaa
bbbbb",
  "highlightLines": [
    0,
  ],
}
`;

exports[`parseLines removes lines correctly 2`] = `
{
  "code": "aaaaa
bbbbb",
  "highlightLines": [
    0,
  ],
}
`;

exports[`parseLines removes lines correctly 3`] = `
{
  "code": "aaaaa
bbbbbbb
bbbbb",
  "highlightLines": [
    0,
    2,
    0,
    1,
  ],
}
`;

exports[`parseLines respects language 1`] = `
{
  "code": "# highlight-next-line
aaaaa
bbbbb",
  "highlightLines": [],
}
`;

exports[`parseLines respects language 2`] = `
{
  "code": "/* highlight-next-line */
aaaaa
bbbbb",
  "highlightLines": [],
}
`;

exports[`parseLines respects language 3`] = `
{
  "code": "// highlight-next-line
aaaa
/* highlight-next-line */
bbbbb
ccccc
<!-- highlight-next-line -->
dddd",
  "highlightLines": [
    4,
  ],
}
`;

exports[`parseLines respects language 4`] = `
{
  "code": "aaaa
bbbbb
ccccc
dddd",
  "highlightLines": [
    0,
    1,
    2,
    3,
  ],
}
`;
